omnipaxos_storage/src/persistent_storage.rs

```diff
@@ -158,127 +269,152 @@ where
-        // Required check because Commitlog has a bug where appending an empty set of entries will
-        // always return an offset.first() with 0 despite entries being in the log.
-        if entries.is_empty() {
-            return self.get_log_len();
-        }
-        let mut serialized = vec![];
+    fn append_entries(&mut self, entries: Vec<T>) -> StorageResult<()> {
+        let mut batch = WriteBatchWithTransaction::<false>::default();
         for entry in entries {
-            serialized.push(bincode::serialize(&entry)?)
+            batch.put_cf(
+                self.get_log_handle(),
+                self.next_log_key.to_be_bytes(),
+                bincode::serialize(&entry)?,
+            );
+            self.next_log_key += 1;
         }
-        let offset = self
-            .commitlog
-            .append(&mut MessageBuf::from_iter(serialized))?;
-        self.commitlog.flush()?; // ensure durable writes
-        Ok(offset.first() + offset.len() as u64)
+        self.db.write(batch)?;
+        Ok(())
     }
 
-    fn append_on_prefix(&mut self, from_idx: u64, entries: Vec<T>) -> StorageResult<u64> {
-        if from_idx > 0 && from_idx < self.get_log_len()? {
-            self.commitlog.truncate(from_idx)?;
+    fn append_on_prefix(&mut self, from_idx: usize, entries: Vec<T>) -> StorageResult<()> {
+        if from_idx < self.next_log_key {
+            let from_key = from_idx.to_be_bytes();
+            let to_key = self.next_log_key.to_be_bytes();
+            self.db
+                .delete_range_cf(self.get_log_handle(), from_key, to_key)?;
+            self.next_log_key = from_idx;
         }
         self.append_entries(entries)
     }
 
-    fn get_entries(&self, from: u64, to: u64) -> StorageResult<Vec<T>> {
-        // Check if the commit log has entries up to the requested endpoint.
-        if to > self.commitlog.next_offset() || from >= to {
+    fn get_entries(&self, mut from: usize, to: usize) -> StorageResult<Vec<T>> {
+        // Check if the log has entries up to the requested endpoint.
+        if to > self.next_log_key || from >= to {
             return Ok(vec![]); // Do an early return
         }
 
-        let buffer = self.commitlog.read(from, ReadLimit::default())?;
-        let mut entries = Vec::<T>::with_capacity((to - from) as usize);
-        let mut iter = buffer.iter();
-        for _ in from..to {
-            let msg = iter.next().ok_or(ErrHelper {})?;
-            entries.push(bincode::deserialize(msg.payload())?);
+        let mut iter = self.db.raw_iterator_cf(self.get_log_handle());
+        let mut entries = Vec::with_capacity(to - from);
+        iter.seek(from.to_be_bytes());
+        loop {
+            let entry_bytes = iter.value().ok_or(ErrHelper {})?;
+            entries.push(bincode::deserialize(entry_bytes)?);
+            from += 1;
+            if from == to {
+                break;
+            }
+            iter.next();
```

Comment:
I prefer using for loops instead of while loops when possible to future-proof against bugs with looping infinitely.